src/amp-story-player/amp-story-player-impl.js

```diff
@@ -870,13 +891,52 @@ export class AmpStoryPlayer {
    * @param {!Element} story
    * @param {!Element} iframe
    * @param {!VisibilityState} visibilityState
+   * @return {!Promise}
    * @private
    */
   layoutIframe_(story, iframe, visibilityState) {
-    this.maybeGetCacheUrl_(story.href).then((url) => {
-      const {href} = this.getEncodedLocation_(url, visibilityState);
-      iframe.setAttribute('src', href);
-      iframe.setAttribute('title', story.textContent.trim());
+    return this.hasSameHref_(story.href, iframe.src).then((hasBeenLaidOut) => {
+      if (hasBeenLaidOut) {
+        return Promise.resolve();
+      }
+
+      let waitPromise;
```

Comment:
When reassigning the `waitPromise` (aka navigation), I think it'd be great to reject the previous `waitPromise`. This way we know for sure that the success callbacks you registered will never trigger.

It'd protect ourselves from the following scenario:

1. Wait for active story to load (it takes a while), promise is pending and waiting to set a `src` on the next iframe
2. User swipes next, which loads the new story
3. First story finally loads and the callback triggers, setting the iframe `src` again

It's fine in this scenario (as long as it doesn't reload the iframe) but if users start adding stories through the programmatic API it can get weird.